File: samples/da-trey-research-python/src/services/db_service.py
import json
import os

from azure.data.tables import TableClient, TableServiceClient

from .utilities import HttpError


class DbService:
    """Generic database service for Azure Table Storage operations."""

    def __init__(self, ok_to_cache_locally: bool = False):
        self._connection_string = os.environ.get("STORAGE_ACCOUNT_CONNECTION_STRING", "")
        if not self._connection_string:
            raise RuntimeError("STORAGE_ACCOUNT_CONNECTION_STRING is not set")
        self._ok_to_cache_locally = ok_to_cache_locally
        self._entity_cache: list = []

    def get_entity_by_row_key(self, table_name: str, row_key: str) -> dict:
        if not self._ok_to_cache_locally:
            table_client = TableClient.from_connection_string(self._connection_string, table_name)
            entity = table_client.get_entity(partition_key=table_name, row_key=row_key)
            return self._expand_property_values(entity)
        else:
            entities = self.get_entities(table_name)
            result = [e for e in entities if e.get("rowKey") == row_key or e.get("RowKey") == row_key]
            if not result:
                raise HttpError(404, f"Entity {row_key} not found")
            return result[0]

    def get_entities(self, table_name: str) -> list:
        if not self._ok_to_cache_locally or not self._entity_cache:
            table_client = TableClient.from_connection_string(self._connection_string, table_name)
            entities = table_client.list_entities()
            self._entity_cache = []
            seen_keys = set()
            for entity in entities:
                rk = entity.get("RowKey", "")
                if rk not in seen_keys:
                    seen_keys.add(rk)
                    self._entity_cache.append(self._expand_property_values(entity))
        return self._entity_cache

    def create_entity(self, table_name: str, row_key: str, entity: dict) -> None:
        self._entity_cache = []
        compressed = self._compress_property_values(entity)
        table_client = TableClient.from_connection_string(self._connection_string, table_name)
        try:
            table_client.create_entity(entity={
                "PartitionKey": table_name,
                "RowKey": row_key,
                **compressed
            })
        except Exception as ex:
            if hasattr(ex, "status_code") and ex.status_code == 409:
                pass  # Entity already exists
            else:
                raise HttpError(500, str(ex))

    def update_entity(self, table_name: str, entity: dict) -> None:
        self._entity_cache = []
        compressed = self._compress_property_values(entity)
        table_client = TableClient.from_connection_string(self._connection_string, table_name)
        table_client.update_entity(entity=compressed, mode="replace")

    def _expand_property_values(self, entity: dict) -> dict:
        result = {}
        for key, value in entity.items():
            result[key] = self._expand_property_value(value)
        return result

    def _expand_property_value(self, value):
        if isinstance(value, str) and len(value) > 0 and value[0] in ('{', '['):
            try:
                return json.loads(value)
            except (json.JSONDecodeError, ValueError):
                return value
        return value

    def _compress_property_values(self, entity: dict) -> dict:
        result = {}
        for key, value in entity.items():
            result[key] = self._compress_property_value(value)
        return result

    def _compress_property_value(self, value):
        if isinstance(value, (dict, list)):
            return json.dumps(value)
        return value

File: samples/da-trey-research-auth-csharp/TreyResearchApi/Services/DbService.cs
using Azure.Data.Tables;
using System.Text.Json;
using TreyResearchApi.Models;

namespace TreyResearchApi.Services;

public class DbService
{
    private readonly string _connectionString;
    private readonly bool _okToCacheLocally;
    private Dictionary<string, object>? _entityCache;

    public DbService(bool okToCacheLocally)
    {
        _connectionString = Environment.GetEnvironmentVariable("STORAGE_ACCOUNT_CONNECTION_STRING")
            ?? throw new InvalidOperationException("STORAGE_ACCOUNT_CONNECTION_STRING is not set");
        _okToCacheLocally = okToCacheLocally;
    }

    public async Task<Dictionary<string, object>?> GetEntityByRowKeyAsync(string tableName, string rowKey)
    {
        if (!_okToCacheLocally)
        {
            var tableClient = new TableClient(_connectionString, tableName);
            var response = await tableClient.GetEntityAsync<TableEntity>(tableName, rowKey);
            return ExpandPropertyValues(response.Value);
        }

        var entities = await GetEntitiesAsync(tableName);
        var result = entities.FirstOrDefault(e =>
            e.TryGetValue("RowKey", out var rk) && rk?.ToString() == rowKey);

        if (result == null)
        {
            throw new HttpError(404, $"Entity {rowKey} not found");
        }

        return result;
    }

    public async Task<List<Dictionary<string, object>>> GetEntitiesAsync(string tableName)
    {
        if (_okToCacheLocally && _entityCache != null)
        {
            return (List<Dictionary<string, object>>)_entityCache["entities"];
        }

        var tableClient = new TableClient(_connectionString, tableName);
        var entities = new List<Dictionary<string, object>>();

        await foreach (var entity in tableClient.QueryAsync<TableEntity>())
        {
            // Remove duplicates
            if (!entities.Any(e => e.TryGetValue("RowKey", out var rk) && rk?.ToString() == entity.RowKey))
            {
                entities.Add(ExpandPropertyValues(entity));
            }
        }

        if (_okToCacheLocally)
        {
            _entityCache = new Dictionary<string, object> { ["entities"] = entities };
        }

        return entities;
    }

    public async Task CreateEntityAsync(string tableName, string rowKey, Dictionary<string, object> entity)
    {
        _entityCache = null;
        var tableEntity = CompressPropertyValues(entity);
        tableEntity["PartitionKey"] = tableName;
        tableEntity["RowKey"] = rowKey;

        var tableClient = new TableClient(_connectionString, tableName);
        var te = new TableEntity(tableEntity.ToDictionary(kvp => kvp.Key, kvp => kvp.Value));

        try
        {
            await tableClient.AddEntityAsync(te);
        }
        catch (Azure.RequestFailedException ex) when (ex.Status == 409)
        {
            // Entity already exists, ignore
        }
        catch (Exception ex)
        {
            throw new HttpError(500, ex.Message);
        }
    }

    public async Task UpdateEntityAsync(string tableName, Dictionary<string, object> entity)
    {
        _entityCache = null;
        var compressed = CompressPropertyValues(entity);
        var te = new TableEntity(compressed.ToDictionary(kvp => kvp.Key, kvp => kvp.Value));

        var tableClient = new TableClient(_connectionString, tableName);
        await tableClient.UpdateEntityAsync(te, Azure.ETag.All, TableUpdateMode.Replace);
    }

    private Dictionary<string, object> ExpandPropertyValues(TableEntity entity)
    {
        var result = new Dictionary<string, object>();
        foreach (var kvp in entity)
        {
            result[kvp.Key] = ExpandPropertyValue(kvp.Value);
        }
        // Ensure standard keys are present
        result["PartitionKey"] = entity.PartitionKey;
        result["RowKey"] = entity.RowKey;
        return result;
    }

    private object ExpandPropertyValue(object value)
    {
        if (value is string s && s.Length > 0 && (s[0] == '{' || s[0] == '['))
        {
            try
            {
                return JsonSerializer.Deserialize<JsonElement>(s);
            }
            catch
            {
                return value;
            }
        }
        return value;
    }

    private Dictionary<string, object> CompressPropertyValues(Dictionary<string, object> entity)
    {
        var result = new Dictionary<string, object>();
        foreach (var kvp in entity)
        {
            result[kvp.Key] = CompressPropertyValue(kvp.Value);
        }
        return result;
    }

    private object CompressPropertyValue(object value)
    {
        if (value is JsonElement je)
        {
            return je.ToString();
        }
        if (value is List<string> list)
        {
            return JsonSerializer.Serialize(list);
        }
        if (value is List<HoursEntry> hours)
        {
            return JsonSerializer.Serialize(hours);
        }
        if (value is Location loc)
        {
            return JsonSerializer.Serialize(loc);
        }
        return value;
    }
}
